funflow-tutorial/notebooks/CustomTaskAndConfigurable.ipynb

```diff
@@ -0,0 +1,310 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    ":opt no-lint\n",
+    "-- Required language extensions\n",
+    "{-# LANGUAGE DataKinds #-}             -- typelevel lists and tuples\n",
+    "{-# LANGUAGE GADTs #-}                 -- task declaration\n",
+    "{-# LANGUAGE OverloadedLabels #-}      -- named task disambiguation\n",
+    "{-# LANGUAGE OverloadedStrings #-}     -- flexible string types\n",
+    "{-# LANGUAGE RankNTypes #-}            -- polymorphism of Flow\n",
+    "\n",
+    "-- core types and functions used here\n",
+    "import Funflow.Config (render, Configurable (..), ExternalConfig)\n",
+    "\n",
+    "-- other necessary imports\n",
+    "import System.Environment (setEnv, unsetEnv)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Custom tasks, `Configurable`s, and type checking\n",
+    "\n",
+    "In this tutorial, we'll show how to build a `Flow` that uses as custom task type.\n",
+    "Additionally, we'll demonstrate `funflow`'s type safety for `Configurable` parameters."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Defining a custom task type and its _interpreter_\n",
+    "\n",
+    "Here, we declare a custom task type and then define __what it does__ with an _interpreter_ function."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import Control.Arrow (Arrow, arr)\n",
+    "\n",
+    "-- Define the representation of a custom task with some String and Int parameters.\n",
+    "-- We'll use this as a sort of closure, yielding a \"function\" (task) with text input and output.\n",
+    "-- This is just the declaration; the closure-like logic lies in the implementation of this task's interpreter.\n",
+    "data EchoXTask i o where\n",
+    "    EchoXTask :: String -> Configurable Int -> EchoXTask String String\n",
+    "\n",
+    "-- our custom task's interpreter; essentially, this is what the task does.\n",
+    "interpretEcho :: (Arrow a) => ExternalConfig -> EchoXTask i o -> a i o\n",
```

Comment:
As we were chatting about on Friday, a cleaner type signature for this would be to use the reading / writing interface in kernmantle to do something like:

```haskell
interpretEcho :: (Arrow core) => EchoXTask i o -> (Writer ConfigKeysBySource ~> Reader ExternalConfig ~> core) i o
```

where your implementation would roughly look like:

```haskell
interpretEcho task = 
  let requiredConfigs = ... in 
  writing requiredConfigs $ 
    reading $ \externalConfig -> 
      render task externalConfig ...
```

Also note that if you want to access IO in the core effect your interpreter returns, you'll need to add an additional constraint of `HasKleisliIO m core`